compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
         }
     }
 
+    /// Checks if an expression refers to a function marked with
+    /// `#[rustc_legacy_const_generics]` and returns the argument index list
+    /// from the attribute.
+    fn legacy_const_generic_args(&mut self, expr: &Expr) -> Option<Vec<usize>> {
+        if let ExprKind::Path(None, path) = &expr.kind {
+            if path.segments.last().unwrap().args.is_some() {
+                return None;
+            }
+            if let Some(partial_res) = self.resolver.get_partial_res(expr.id) {
+                if partial_res.unresolved_segments() != 0 {
+                    return None;
+                }
+                if let Res::Def(hir::def::DefKind::Fn, def_id) = partial_res.base_res() {
+                    let attrs = self.item_attrs(def_id);
+                    let attr = attrs
+                        .iter()
+                        .find(|a| self.sess.check_name(a, sym::rustc_legacy_const_generics))?;
+                    let mut ret = vec![];
+                    for meta in attr.meta_item_list()? {
+                        match meta.literal()?.kind {
+                            LitKind::Int(a, _) => {
+                                ret.push(a as usize);
+                            }
+                            _ => panic!("invalid arg index"),
+                        }
+                    }
+                    return Some(ret);
+                }
+            }
+        }
+        None
+    }
+
+    fn lower_legacy_const_generics(
+        &mut self,
+        mut f: Expr,
+        args: Vec<AstP<Expr>>,
+        legacy_args_idx: &[usize],
+    ) -> hir::ExprKind<'hir> {
+        let path = match f.kind {
+            ExprKind::Path(None, ref mut path) => path,
+            _ => unreachable!(),
+        };
+
+        // Split the arguments into const generics and normal arguments
+        let mut real_args = vec![];
+        let mut generic_args = vec![];
+        for (idx, arg) in args.into_iter().enumerate() {
+            if legacy_args_idx.contains(&idx) {
+                let parent_def_id = self.current_hir_id_owner.last().unwrap().0;
+                let node_id = self.resolver.next_node_id();
+
+                // Add a definition for the in-band const def.
+                self.resolver.create_def(
+                    parent_def_id,
+                    node_id,
+                    DefPathData::AnonConst,
+                    ExpnId::root(),
+                    arg.span,
+                );
+
+                let anon_const = AnonConst { id: node_id, value: arg };
```

Comment:
Could you add a test for the case where the argument is not a valid constant (e.g. `let a = 1; foo(0, a, 0)`)?

Reply:
The compiler ICEs in this case: late resolution happens before AST lowering and needs to be aware of the rewrite since the expression in the `AnonConst` needs to be resolved inside a `ConstantItemRibKind`.

I will fix this and add a test case.